### PR TITLE
An initial draft of updating recomendations

### PR DESCRIPTION
**Pull request by MovieStoreGuy**
In order to help making sub packages across the repo more consistent,
the contribution guide is being updated to reflect that.

**Description:**
Looking to expand contributions to make consistent approaches to working on the collector and the contrib.

**Link to tracking Issue:**
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/6209

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -289,6 +299,32 @@ do not decrease overall code coverage of the codebase - this is aligned with our
...
+#### Library Recommendations 
+
+To keep testing practices consistent across the project, it is advised to use this libraries under
+these circumstances:
+
+- For assertions to validate expecations, use `"github.com/stretchr/testify/assert"`
+- For assertions that are required to continue the test, use `"github.com/stretchr/testify/require"`
+- For mocking external resources, use `"github.com/stretchr/testify/mock"`
+- For validating HTTP traffic interactions, `"net/http/httptest"`
+- For generating test data, use `???`
```
Comment (MovieStoreGuy) [inline CONTRIBUTING.md:311]: I have a feeling that there is or should be a library to do this as part of some testing framework.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -289,6 +299,32 @@ do not decrease overall code coverage of the codebase - this is aligned with our
...
+
+### Integration Testing
+
+Integration testing is encourage throughout the project, docker images can be used in order to facility
+a local version. The recommend library to control docker images throughout testing is `???`.
```
Comment (MovieStoreGuy) [inline CONTRIBUTING.md:316]: I believe I have seem packages use a docker controller to manage which would be good for integration tests.

**Review by MovieStoreGuy — COMMENTED** [review]
Just adding some of the thoughts

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -170,6 +170,16 @@ for coding advice). The code must adhere to the following robustness principles
 are important for software that runs autonomously and continuously without direct
 interaction with a human (such as this Collector).
 
+### Recommended Libraries / Defaults
+
+In order to simplifiy developing within the project, library recommendations have been set
+and should be followed.
```
Comment (jpkrohling) [inline CONTRIBUTING.md:176]: Add also something like, "There are places in the codebase that are still not following this, but new code should strive to align with the recommendations."

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -289,6 +299,32 @@ do not decrease overall code coverage of the codebase - this is aligned with our
 goal to increase coverage over time. Keep track of execution time for your unit
 tests and try to keep them as short as possible.
 
+#### Library Recommendations 
+
+To keep testing practices consistent across the project, it is advised to use this libraries under
```
Comment (jpkrohling) [inline CONTRIBUTING.md:304]:
```suggestion
To keep testing practices consistent across the project, it is advised to use these libraries under
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -289,6 +299,32 @@ do not decrease overall code coverage of the codebase - this is aligned with our
...
+these circumstances:
+
+- For assertions to validate expecations, use `"github.com/stretchr/testify/assert"`
```
Comment (jpkrohling) [inline CONTRIBUTING.md:307]:
```suggestion
- For assertions to validate expectations, use `"github.com/stretchr/testify/assert"`
```

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -289,6 +299,32 @@ do not decrease overall code coverage of the codebase - this is aligned with our
...
+- For assertions that are required to continue the test, use `"github.com/stretchr/testify/require"`
+- For mocking external resources, use `"github.com/stretchr/testify/mock"`
+- For validating HTTP traffic interactions, `"net/http/httptest"`
+- For generating test data, use `???`
+
+### Integration Testing
+
+Integration testing is encourage throughout the project, docker images can be used in order to facility
```
Comment (jpkrohling) [inline CONTRIBUTING.md:315]:
```suggestion
Integration testing is encouraged throughout the project, container images can be used in order to facilitate
```

Not sure I get the part about the container images.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -170,6 +170,16 @@ for coding advice). The code must adhere to the following robustness principles
...
 interaction with a human (such as this Collector).
 
+### Recommended Libraries / Defaults
+
+In order to simplifiy developing within the project, library recommendations have been set
+and should be followed.
+
+Scenario | Recomended                                   | Rationale 
+----------------------------------------------------------------------------
+Hashing  | ["hashing/fnv"](https://pkg.go.dev/hash/fnv) | The project adopted this as the default hashing method due to the efficency and is reasonable for non cryptographic use
```
Comment (tigrannajaryan) [inline CONTRIBUTING.md:180]: Why is this not in the Library Recommendations section? Is this somehow special?

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -289,6 +299,32 @@ do not decrease overall code coverage of the codebase - this is aligned with our
...
+
+### Integration Testing
+
+Integration testing is encourage throughout the project, docker images can be used in order to facility
+a local version. The recommend library to control docker images throughout testing is `???`.
+In the absence of locally running version, mocking should be used to assert expectations.
+
+### Using CGO
+
+Using CGO should be avoided due to the lack of portability and complexity
```
Comment (tigrannajaryan) [inline CONTRIBUTING.md:321]:
```suggestion
Using CGO is prohibited due to the lack of portability and complexity 
```

Comment (tigrannajaryan) [reply CONTRIBUTING.md:321]: As far as I know we don't allow CGO at all currently. Not sure if there is a need to carve out exceptions. If the component is never going to be included in the build here it is best to make it clear upfront.

Comment (MovieStoreGuy) [reply CONTRIBUTING.md:321]: Is that only true for the collector? or does that include the contrib?

File: CONTRIBUTING.md (hunk 7 above)
Comment (MovieStoreGuy) [reply CONTRIBUTING.md:180]: The added recommended lib sections that I had added was specific to testing but I am happy to shift it into one spot if it makes more sense :)

File: CONTRIBUTING.md (hunk 6 above)
Comment (MovieStoreGuy) [reply CONTRIBUTING.md:315]: I think this relates more to the contrib collector where in exporter's test (and potentially receivers), containers of the offical thing has been started up in their test and controlled by it. (I feel like I have seen this, also high potential I had dreamt it).

File: CONTRIBUTING.md (hunk 3 above)
Comment (MovieStoreGuy) [reply CONTRIBUTING.md:176]: 👍🏽  ,

Yeah, I agree with that.

File: CONTRIBUTING.md (hunk 8 above)
Comment (mx-psi) [reply CONTRIBUTING.md:323]: We use cgo optionally on the hostmetricsreceiver macOS implementation (see [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/hostmetricsreceiver#notes)), I think that is the only place we use it today

Comment (MovieStoreGuy) [reply CONTRIBUTING.md:323]: Alright, sounds good to me :)

**Comment by MovieStoreGuy** [comment]
Keen to hear your thoughts on this and if there is anything immediate that show be extended.

File: CHANGELOG.md [hunk 9]
```diff
@@ -6,6 +6,9 @@
 
 - Fix handling of corrupted records by persistent buffer (experimental) (#4475)
 
+## 💡 Enhancements 💡
+
+- Extending the contribution guide to help clarify what is acceptable defaults and recommendations.
```
Comment (jpkrohling) [inline CHANGELOG.md:11]: Another point to the recommendations doc: when should a changelog entry be added?

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -289,6 +301,40 @@ do not decrease overall code coverage of the codebase - this is aligned with our
...
+Using CGO is prohibited due to the lack of portability and complexity 
+that comes with managing external libaries with different operating systems and configurations.
+However, if the package MUST use CGO, this should be explicitly called out within the readme
+with clear instructions on how to install the required libraries.
+Furthermore, if your package requires CGO, it MUST be able to compile and operate in a no op mode
+or report a warning back to the collector with a clear error saying CGO is required to work.
+
+### Configuration Naming
+
+In order to keep configuration settings consistent across the project and their intenet,
```
Comment (jpkrohling) [inline CONTRIBUTING.md:330]:
```suggestion
In order to keep configuration settings consistent across the project and their intent,
```

**Review by jpkrohling — COMMENTED** [review]
Other than a typo, this looks like a great addition.

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -289,6 +301,40 @@ do not decrease overall code coverage of the codebase - this is aligned with our
...
+a package must use the following naming for their purpose:
+
+| Name               | Purpose                                                              | Example                                                     |
+|--------------------|----------------------------------------------------------------------|-------------------------------------------------------------|
+| ExportTimeout      | Used to set the allowed during of an export operation                | Setting the allowed time a single export operation can take |
```
Comment (jpkrohling) [inline CONTRIBUTING.md:335]: This seems to be in response to #4497, which was discussed yesterday during the SIG meeting. Let's wait for the resolution before merging this.

Comment (MovieStoreGuy) [reply CONTRIBUTING.md:335]: Yes, sorry I was meant to ping you on it but it had slipped my mind.

File: CHANGELOG.md (hunk 9 above)
Comment (MovieStoreGuy) [reply CHANGELOG.md:11]: I think what I have in mind for this:

- Behaviourally changes
- Fixes
- Breaking changes

I things like CI changes, testing updates, internal helper package changes (provided it doesn't fall into the above three), don't really need to be consumed by a user of the collector.

Did you have something similar in mind?

**Comment by MovieStoreGuy** [comment]
I do have a question on, should we encourage `t.Parallel()` when writing tests?

File: CHANGELOG.md (hunk 9 above)
Comment (jpkrohling) [reply CHANGELOG.md:11]: Would the changes to the contributing guide fit this? We have two main audiences for this repository: 

- component developers, interested in the API changes we introduce
- end-users, interested in the config changes for the components located here (otlpexporter and receiver, for instance)

IMO, we should have entries only for changes affecting those audiences.

**Comment by MovieStoreGuy** [comment]
Hey @jpkrohling,

Let me know of your thoughts thus far on this :) 

I am considering leaving any additions out for a future PR and allow this to be an iterative process.

File: CONTRIBUTING.md (hunk 10 above)
Comment (jpkrohling) [reply CONTRIBUTING.md:332]: This doesn't seem to have been fixed.

**Review by jpkrohling — COMMENTED** [review]
Other than a typo, this looks like a great start.

**Comment by MovieStoreGuy** [comment]
Sorry, that should be addressed now @jpkrohling :) 

I had squashed the commits since they were becoming rather "fixing lint" and weren't helpful.

File: CONTRIBUTING.md (hunk 11 above)
Comment (jpkrohling) [reply CONTRIBUTING.md:337]: In order to clear this PR for merging, would you please remove this, or add a link to the discussion? This doesn't seem to be settled yet.

Comment (MovieStoreGuy) [reply CONTRIBUTING.md]: Done :)

**Review by jpkrohling — APPROVED** [review]

**Comment by tigrannajaryan** [comment]
Thanks @MovieStoreGuy

**Pull request merged**